### PR TITLE
[Merged by Bors] - feat: port CategoryTheory.Functor.Flat

### PR DESCRIPTION
**Pull request by kim-em**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by kbuzzard** [comment]
Right now this is blocked by https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/change.20in.20behaviour.20with.20.60simps.60/near/354350606 ; a `simp` proof broke and I've not been able to fix it yet, although I've diagnosed the issue.

File: Mathlib/CategoryTheory/Functor/Flat.lean [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+variable {C : Type u₁} [Category.{v₁} C] {D : Type u₂} [Category.{v₁} D]
+
+variable {J : Type w} [SmallCategory J]
+
+variable {K : J ⥤ C} (F : C ⥤ D) (c : Cone K)
+
+-- **TODO** Scott changed `@[simps]` to `@[simps!]` below and I don't
+-- know what this does, but one thing it does is that
+-- it stops `toDiagram_obj` being created, and `toDiagram_obj` is
+-- used later on so I (kmb) have changed it back
```
Comment (kbuzzard) [inline Mathlib/CategoryTheory/Functor/Flat.lean:70]: @semorrison  In [this commit](https://github.com/leanprover-community/mathlib4/pull/3703/commits/212f7539353e0a8327b292354cfd2676488f2038) in the port of CategoryTheory.Functor.Flat you changed the `@[simps]` on `toDiagram` to `@[simps!]`. This breaks the proof of `uniq` (see line 249 or so). I don't know what `@[simps!]` does (or how to find out) but I thought I'd flag this.

Comment (semorrison) [reply Mathlib/CategoryTheory/Functor/Flat.lean:70]: Thanks, `simps!` was wrong. It's for cases where `simps` needs to unfold a definition before it finds a constructor (like `{rhs_md := reducible}` in mathlib3). Here it's just a constructor, sorry I don't know why I changed it. 🤷‍♀️

**Comment by semorrison** [comment]
@kbuzzard, I haven't looked at this one recently (apologies for nabbing it, I shouldn't have). I'll let you change the labels if/when it's ready for review, or just ping me?

**Comment by kbuzzard** [comment]
OK I'm happy for this to be reviewed now. Here are the remaining issues which perhaps merit discussion: 

1) line 67 @semorrison you changed `@[simps]` to `@[simps!]` and I don't know what this does (I couldn't find any documentation on it) but it broke a proof so I changed it back. More details are in the file: see the two **TODO**s. Scott if you're happy to go with `@[simps]` then these TODOs can be deleted, I just wanted to flag them because I didn't understand what was happpening.

2) Line 374 there's a failing `ext` which I replaced with a manual application of the relevant `ext` lemma. I think that writing the bespoke `ext` lemma here is a bit farcical: under the hypotheses of the theorem, ` ((lan F ⋙ (evaluation D E).obj X).obj G` is defeq to `colimit (Lan.diagram F G X)`, so it would be a super-niche super-weird ext lemma. 

3) The instances on lines 431 to 433 are probably in the wrong place; I left them here for now because I wasn't sure if we wanted them at all or if we wanted another hack instead.

4) Twice I had to stop in the middle of a `simp` proof and rewrite `Comma.eqToHom_right` manually. I've tried really hard to minimise (several hours) but I can't. Maybe I'm missing something. I asked at https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/simp.20not.20using.20a.20simp.20lemma/near/353943416 but because I can't minimise the question isn't really a very good one.

5) The type of `lan_map_app` has changed definitionally; this is not a declaration added by the PR so is not really this PR's problem; I asked at https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/change.20in.20behaviour.20with.20.60simps.60/near/354350606


Issue (1) above should be easily to deal with  once I'm educated in `simp!`, issue 3 should also be easy to deal with, issues (2) and (4) and (5) can just be ignored I guess if we just want to get on with things. Nobody responded to my Zulip questions for (4) and (5).

File: Mathlib/CategoryTheory/Limits/Preserves/FunctorCategory.lean [hunk 2]
```diff
@@ -4,7 +4,7 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Bhavik Mehta
 
 ! This file was ported from Lean 3 source module category_theory.limits.preserves.functor_category
-! leanprover-community/mathlib commit 7cd8adb7a9d7d0498d2e76c23cd4255f966899f5
+! leanprover-community/mathlib commit 39478763114722f0ec7613cb2f3f7701f9b86c8d
```
Comment (eric-wieser) [inline Mathlib/CategoryTheory/Limits/Preserves/FunctorCategory.lean:7]: I've moved this to #3798

Comment (kbuzzard) [reply Mathlib/CategoryTheory/Limits/Preserves/FunctorCategory.lean]: I've manually reverted the changes to this file in this PR.

**Comment by kbuzzard** [comment]
I asked about 3 here https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/typeclass.20inference.20not.20seeing.20through.20reducible.20definition/near/356130343

**Comment by semorrison** [comment]
@kbuzzard, `@[simps!]` is approximately the replacement for mathlib3's `rhs_md := reducible`. If it's better to not use it here, completely fine to revert.

**Comment by kbuzzard** [comment]
Re 3: Adam Topaz has suggested unification hints. Does anyone know how these work? Right now https://leanprover.github.io/lean4/doc/unifhint.html is empty.

**Comment by kbuzzard** [comment]
OK I think that the remaining issues are just porting notes or different PRs, so I'm not going to work any more on this PR. I've tidied up and it's ready for review.

File: Mathlib/CategoryTheory/ConcreteCategory/Basic.lean [hunk 3]
```diff
@@ -75,6 +75,13 @@ instance ConcreteCategory.types : ConcreteCategory (Type u) where
   Forget := 𝟭 _
 #align category_theory.concrete_category.types CategoryTheory.ConcreteCategory.types
 
+-- porting note: these were all inferred in mathlib3 because lean 3 typeclass inference could see
+-- that `forget (Type u) = 𝟭 (Type u)`
+-- see https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/typeclass.20inference.20failure/near/354843721
+instance : ReflectsLimits (forget (Type u₁)) := Limits.idReflectsLimits
+instance : PreservesColimits (forget (Type u₁)) := Limits.idPreservesColimits
+instance : PreservesLimits (forget (Type u₁)) := Limits.idPreservesLimits
```
Comment (eric-wieser) [inline Mathlib/CategoryTheory/ConcreteCategory/Basic.lean:83]: I think as discussed in that thread, we should just add `@[reducible]` instance in some places.

Comment (eric-wieser) [reply Mathlib/CategoryTheory/ConcreteCategory/Basic.lean:83]: Let's see if #3857 fixes this before committing these

**Review by eric-wieser — CHANGES_REQUESTED** [review]

**Comment by semorrison** [comment]
I merged master to incorporate #3857.

File: Mathlib/CategoryTheory/ConcreteCategory/Basic.lean (hunk 3 above)
Comment (kbuzzard) [reply Mathlib/CategoryTheory/ConcreteCategory/Basic.lean]: It did.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**